### PR TITLE
Add client-side handling for server-side errors using `errorComponent` and `wrapInSuspense: true`

### PR DESCRIPTION
**Pull request by giilbert**
Addresses issue #968 fixing error handling during SSR

This pull request adds:
- The ability for server errors to be handled via `errorComponent` while `wrapInSuspense: true`
- Router options `serializeError` and `deserializeError` (with sensible defaults) that allows an end-user to decide what parts of an error to forward from the server to the client
- `/error` pages on the SSR examples

### CODE REVIEW

**Comment by giilbert** [comment]
This fixes error handling... but only for non-`defer`ed promises. I just noticed that it doesn't fix it for `defer`ed promises. Looking into it now... :sob:

**Comment by giilbert** [comment]
Error handling in (streaming) SSR seems to work! but the API and documentation need some more work. I'm not 200% sure if I did everything correctly - please give me some feedback, I'm gonna call it a day.

**Comment by SamiHonkasalo** [comment]
Should the `ResolvedSuspenseBoundary` automatically resolve to a `<Suspense>`, if the route has has a loader and it's being rendered on the server, instead of manually setting `wrapInSuspense: true`?

I tried looking into the way that Remix handles their ErrorBoundary on SSR, but I could not figure it out myself

File: examples/react/basic-ssr-streaming-file-based/src/routes/posts/$postId.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { FileRoute } from '@tanstack/react-router'
+import { CatchBoundary, FileRoute } from '@tanstack/react-router'
```
Comment (schiller-manuel) [inline examples/react/basic-ssr-streaming-file-based/src/routes/posts/$postId.tsx:1]: is this import needed?

File: packages/react-router/src/Matches.tsx [hunk 2]
```diff
@@ -422,3 +428,20 @@ export function useLoaderData<
...
+} {
+  if (!(typeof error === 'object' && error && 'data' in error)) return false
+  if (!('__isServerError' in error && error.__isServerError)) return false
+  if (!(typeof error.data === 'object' && error.data)) return false
+
+  return error.__isServerError === true
+}
+
+export function defaultDeserializeError(serializedData: Record<string, any>) {
+  const error = new Error(serializedData.message)
```
Comment (schiller-manuel) [inline packages/react-router/src/Matches.tsx:444]: what happens if `message` or `name` are not part of `serializedData`?

File: packages/react-router/src/router.ts [hunk 3]
```diff
@@ -125,6 +125,8 @@ export interface RouterOptions<
   Wrap?: (props: { children: any }) => JSX.Element
   InnerWrap?: (props: { children: any }) => JSX.Element
   notFoundRoute?: AnyRoute
+  serializeError?: (err: unknown) => Record<string, any>
```
Comment (schiller-manuel) [inline packages/react-router/src/router.ts:128]: please document those new properties in docs/api/router/RouterOptionsType.md

**Review by schiller-manuel — CHANGES_REQUESTED** [review]

File: docs/api/router/RouterOptionsType.md [hunk 4]
```diff
@@ -179,6 +179,16 @@ const router = new Router({
 - Optional
 - A route that will be used as the default not found route for every branch of the route tree. This can be overridden on a per-branch basis by providing a not found route to the `NotFoundRoute` option on the root route of the branch.
 
+#### `serializeError`
+
+- Type: `(err: unknown) => Record<string, any>`
+- Optional
+- A function that is called to define how errors are serialized when they are stored in the router's dehydrated state.
+
+#### `deserializeError`
```
Comment (schiller-manuel) [inline docs/api/router/RouterOptionsType.md:188]: is this not optional?

Comment (iDarkLightning) [reply docs/api/router/RouterOptionsType.md:188]: oops

**Comment by schiller-manuel** [comment]
please rebase so that we can merge this

**Pull request merged**